File: modules/wem-api/src/main/java/com/enonic/wem/api/content/page/DescriptorKey.java
package com.enonic.wem.api.content.page;

import java.util.Objects;

import com.enonic.wem.api.module.ModuleKey;
import com.enonic.wem.api.module.ModuleResourceKey;

public abstract class DescriptorKey
{
    public enum DescriptorType
    {
        IMAGE,
        LAYOUT,
        PAGE,
        PART
    }

    protected static final String SEPARATOR = ":";

    private final ModuleKey moduleKey;

    private final ComponentDescriptorName name;

    private final String refString;

    private final DescriptorType descriptorType;

    protected DescriptorKey( final ModuleKey moduleKey, final ComponentDescriptorName name, final DescriptorType descriptorType )
    {
        this.descriptorType = descriptorType;
        this.moduleKey = moduleKey;
        this.name = name;
        this.refString = moduleKey.toString() + SEPARATOR + name.toString();
    }

    public ModuleKey getModuleKey()
    {
        return moduleKey;
    }

    public ComponentDescriptorName getName()
    {
        return name;
    }

    public DescriptorType getDescriptorType()
    {
        return descriptorType;
    }

    public abstract ModuleResourceKey toResourceKey();

    @Override
    public boolean equals( final Object o )
    {
        if ( this == o )
        {
            return true;
        }
        if ( o == null || getClass() != o.getClass() )
        {
            return false;
        }

        final DescriptorKey that = (DescriptorKey) o;
        return Objects.equals( this.refString, that.refString );
    }

    @Override
    public int hashCode()
    {
        return Objects.hash( this.refString );
    }

    public String toString()
    {
        return refString;
    }
}

File: modules/wem-webapp/src/main/webapp/admin/common/js/content/page/DescriptorKey.ts
module api.content.page {

    export class DescriptorKey {

        private static SEPARATOR = ":";

        private moduleKey: api.module.ModuleKey;

        private name: ComponentDescriptorName;

        private refString:string;

        public static fromString(str:string):DescriptorKey{
            var sepIndex:number = str.indexOf(DescriptorKey.SEPARATOR);
            if( sepIndex == -1 ) {
                throw new Error("DescriptorKey must contain separator '" + DescriptorKey.SEPARATOR + "':" + str);
            }

            var moduleKey = str.substring(0, sepIndex);
            var name = str.substring(sepIndex+1, str.length);

            return new DescriptorKey(api.module.ModuleKey.fromString(moduleKey), new ComponentDescriptorName(name));
        }

        constructor(moduleKey:api.module.ModuleKey, name: ComponentDescriptorName) {
            this.moduleKey = moduleKey;
            this.name = name;
            this.refString = moduleKey.toString() + DescriptorKey.SEPARATOR  + name.toString();
        }

        getModuleKey():api.module.ModuleKey {
            return this.moduleKey;
        }

        getName():ComponentDescriptorName {
            return this.name;
        }

        toString():string {
            return this.refString;
        }
    }
}